clusterDeployer.py

```diff
@@ -125,27 +128,61 @@ def remove_worker_dhcp_entries(self, hosts: List[ClusterHost]) -> None:
             h.prepare_remove_worker_dhcp_leases(self._local_host.bridge, removed_macs)
         lh = host.LocalHost()
         lh.run("systemctl restart libvirtd")
+        self._local_host.ensure_linked_to_network(self._local_host.bridge)
 
-    def teardown(self) -> None:
+    def teardown_masters(self) -> None:
         cluster_name = self._cc.name
+        if "masters" not in self.steps:
+            logger.info(f"Not tearing down {cluster_name}")
+            return
+
         logger.info(f"Tearing down {cluster_name}")
         self._ai.ensure_cluster_deleted(self._cc.name)
 
-        for h in self._all_hosts:
-            h.teardown()
+        for h in self._all_master_hosts:
+            h.teardown_masters()
 
         self._ai.ensure_infraenv_deleted(f"{cluster_name}-x86_64")
         self._ai.ensure_infraenv_deleted(f"{cluster_name}-arm64")
 
-        self.remove_master_dhcp_entries(self._all_hosts)
-        self.remove_worker_dhcp_entries(self._all_hosts)
+        self.remove_master_dhcp_entries(self._all_master_hosts)
 
         for h in self._all_hosts:
             h.ensure_not_linked_to_network()
+        for h in self._all_master_hosts:
+            h.ensure_not_linked_to_network()
 
         if os.path.exists(self._cc.kubeconfig):
             os.remove(self._cc.kubeconfig)
 
+    def teardown_workers(self) -> None:
+        cluster_name = self._cc.name
+
+        # If workers not in steps (and masters set), teardown the workers to avoid dangling vms.
+        if "workers" in self.steps and "masters" not in self.steps:
+            logger.info(f"Tearing down (some) workers on {cluster_name}")
+        elif "masters" in self.steps:
+            logger.info(f"Tearing down (some) workers on {cluster_name} before tearing down masters")
+        else:
+            return
+
+        for h in self._all_hosts:
+            h.teardown_workers()
+
+        self.remove_worker_dhcp_entries(self._all_hosts)
```

Comment:
We repeat some of these things.  Shouldn't we just make it that the ClusterNode.teardown() method gets a `dhcp_bridge` argument and calls `dhcp_bridge.remove_dhcp_entry`?

Also, can't we compute the list of ClusterNodes that need to be torn down (based on --steps) and then issue their teardown() method?  We don't really need separate teardown_masters/workers methods.

I know that create_masters() and create_workers() are separate but ideally those should be refactored too.  Code duplication is not nice IMO.

Reply:
Next version of PR fixes this, setting up and deleting dhcp entries by calling virtualbridge methods from clusterDeployer.py.